OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs

```diff
@@ -0,0 +1,223 @@
+		{
+			var exit = self.Location + exitInfo.ExitCell;
+			var spawn = self.CenterPosition + exitInfo.SpawnOffset;
+
+			var td = new TypeDictionary
+			{
+				new OwnerInit(self.Owner),
+				new LocationInit(exit),
+				new CenterPositionInit(spawn)
+			};
+
+			if (!string.IsNullOrEmpty(raceVariant))
+				td.Add(new RaceInit(raceVariant));
+
+			return self.World.CreateActor(false, actorToProduce.Name, td);
+		}
+
+		public override bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
+		{
+			var deliveringActorType = info.DeliveryActor;
+			var owner = self.Owner;
+
+			factionVariant = raceVariant;
+
+			// Check if there is a valid drop-off point before sending the transport
+			var exit = GetAvailableExit(self, actorsToProduce.First());
+			if (exit == null)
+				return false;
+
+			foreach (var trait in self.TraitsImplementing<INotifyDelivery>())
+				trait.IncomingDelivery(self);
+
+			owner.World.AddFrameEndTask(w =>
+			{
+				var deliveringActor = w.CreateActor(deliveringActorType, new TypeDictionary
+				{
+					new CenterPositionInit(w.Map.CenterOfCell(startPos) + new WVec(WRange.Zero, WRange.Zero, aircraft.CruiseAltitude)),
+					new OwnerInit(owner),
+					new FacingInit(64)
+				});
+
+				if (isPlane)
+					deliveringActor.QueueActivity(new Fly(deliveringActor, Target.FromCell(w, self.Location + new CVec(9, 0))));
+				else
+					deliveringActor.QueueActivity(new HeliFly(deliveringActor, Target.FromCell(w, self.Location + new CVec(9, 0))));
+
+				deliveringActor.QueueActivity(new CallFunc(() => TryToLand(deliveringActor, actorsToProduce)));
+			});
+
+			return true;
+		}
+
+		void TryToLand(Actor deliveringActor, IEnumerable<ActorInfo> actorsToProduce)
+		{
+			// Check if there is a valid drop-off point before beginning to descend
+			var exit = GetAvailableExit(self, actorsToProduce.First());
+
+			// Abort the landing and refund the player
+			if (exit == null || self.IsDead || !self.IsInWorld)
+			{
+				var value = actorsToProduce.Sum(actor => actor.Traits.Get<ValuedInfo>().Cost);
+				self.Owner.PlayerActor.Trait<PlayerResources>().GiveCash(value);
+
+				if (isPlane)
+					deliveringActor.QueueActivity(new Fly(deliveringActor, Target.FromCell(self.World, endPos)));
+				else
+					deliveringActor.QueueActivity(new HeliFly(deliveringActor, Target.FromCell(self.World, endPos)));
+
+				deliveringActor.QueueActivity(new RemoveSelf());
+
+				return;
+			}
+
+			if (isPlane)
+				deliveringActor.QueueActivity(new Land(deliveringActor, Target.FromActor(self)));
+			else
+			{
+				deliveringActor.QueueActivity(new HeliFly(deliveringActor, Target.FromActor(self)));
+				deliveringActor.QueueActivity(new HeliLand(deliveringActor, false));
+			}
+
+			deliveringActor.QueueActivity(new CallFunc(() => MakeDelivery(actorsToProduce.ToList(), deliveringActor, deliveringActor.World)));
+		}
+
+		void MakeDelivery(ICollection<ActorInfo> actorInfos, Actor deliveringActor, World world)
```

Comment:
It would be clearer to declare `List<ActorInfo> actorInfos` here and in `WaitAndRetry`.